check-list/src/main/java/org/sonar/java/CheckListGenerator.java

```diff
@@ -0,0 +1,271 @@
+
+  final Path relativePath;
+  final Path awsRelativePath;
+  Path pathToWriteList;
+  final String rulesPath;
+
+  public CheckListGenerator(Gson gson, Path relativePath, Path awsRelativePath, Path pathToWriteList, String rulesPath) {
+    this.gson = gson;
+    this.relativePath = relativePath;
+    this.awsRelativePath = awsRelativePath;
+    this.pathToWriteList = pathToWriteList;
+    this.rulesPath = rulesPath;
+  }
+
+  public static void main(String[] args) {
+    var generator = new CheckListGenerator(new Gson(),
+      Path.of("java-checks/src/main/java"),
+      Path.of("java-checks-aws/src/main/java"),
+      Path.of("check-list/target/generated-sources/" + CLASS_NAME + ".java"),
+      RULES_PATH);
+    generateCheckList(generator);
+  }
+
+  public static void generateCheckList(CheckListGenerator generator) {
+    var checks = generator.getCheckClasses(generator.relativePath, generator.awsRelativePath);
+
+    List<Class<?>> mainClasses = new ArrayList<>();
+    List<Class<?>> testClasses = new ArrayList<>();
+    List<Class<?>> allClasses = new ArrayList<>();
+    generator.generateCheckListClasses(checks, mainClasses, testClasses, allClasses, generator.rulesPath);
+    String main = generator.collectChecks(mainClasses);
+    String test = generator.collectChecks(testClasses);
+    String all = generator.collectChecks(allClasses);
+
+    String importChecks = generateImportStatements(checks);
+
+    generator.writeToFile(importChecks, main, test, all, generator.pathToWriteList);
+  }
+
+  public List<? extends Class<?>> getCheckClasses(Path directory, Path awsDirectory) {
+    var modifiedFiles = getModifiedFiles(directory, awsDirectory);
+    return modifiedFiles.stream()
+      .map(CheckListGenerator::getClassByName)
+      .filter(c -> !Modifier.isAbstract(c.getModifiers()))
+      .filter(c -> c.getAnnotationsByType(Rule.class).length != 0)
+      .toList();
+  }
+
+  private static List<String> getModifiedFiles(Path relativePath, Path awsRelativePath) {
+    try (Stream<Path> stream = Stream.concat(Files.walk(relativePath.resolve("org/sonar/java/checks")), Files.walk(awsRelativePath.resolve("org/sonar/java/checks")))) {
+      return stream
+        .map(p -> p.startsWith(awsRelativePath) ? awsRelativePath.relativize(p).toString() : relativePath.relativize(p).toString())
+        .filter(file -> file.endsWith("Check.java"))
+        .map(file -> file.replace(".java", "").replace(File.separator, "."))
+        .toList();
+    } catch (IOException e) {
+      throw new IllegalStateException(e.getMessage());
+    }
+  }
+
+  private static Class<?> getClassByName(String className) {
+    try {
+      return Class.forName(className);
+    } catch (ClassNotFoundException e) {
+      throw new IllegalStateException("Cannot find the class for name " + className, e);
+    }
+  }
+
+  public String getRuleKey(Class<?> check) {
+    return check.getAnnotation(Rule.class).key();
+  }
+
+  public Metadata getMetadata(Reader reader) {
```

Comment:
`Metadata` is protected, but here it is returned by a public method (cf warning in Intellij)